### DISCUSSION TITLE
How to modularize actions?

### DISCUSSION THREAD

**Discussion by ukreddy-erwin** [community/community#26320] [Actions]
<p>There are multiple questions created long back and no response on them later.<br>
The recent one is in 2019 last reply without any solution.<br>
There is no use to resurrect them, so creating a new one.</p>
<p>I have a github action workflow job steps containing 10 to 15 windows batch commands, which I am running in my self-hosted windows server.</p>
<p>The same steps need to be repeated multiple times for different input files.</p>
<p>Is there anyway to create a module of this and call this with the parameters changed?</p>
<p>Please suggest.</p>

**Comment by ViacheslavKudinov** [comment]
<p>Hi <a class="mention" href="/ukreddy-erwin">@ukreddy-erwin</a><br>
There are a few ways to reduce duplication: composite actions and reusable workflows.<br>
They can’t solve all aspects but could help.<br>
Please, take a look at a few articles:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://docs.github.com/en/actions/using-workflows/reusing-workflows">
  <header class="source">
      <img src="https://docs.github.com/assets/cb-803/images/site/favicon.svg" class="site-icon" width="16" height="16">

      <a href="https://docs.github.com/en/actions/using-workflows/reusing-workflows" target="_blank" rel="noopener nofollow ugc">GitHub Docs</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/56436734/181096882-1a144c6a-8e2c-4f0d-9fa9-d84669310f7c.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://docs.github.com/en/actions/using-workflows/reusing-workflows" target="_blank" rel="noopener nofollow ugc">Reusing workflows - GitHub Docs</a></h3>

  <p>Learn how to avoid duplication when creating a workflow by reusing existing workflows.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://docs.github.com/en/actions/creating-actions/creating-a-composite-action">
  <header class="source">
      <img src="https://docs.github.com/assets/cb-803/images/site/favicon.svg" class="site-icon" width="16" height="16">

      <a href="https://docs.github.com/en/actions/creating-actions/creating-a-composite-action" target="_blank" rel="noopener nofollow ugc">GitHub Docs</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/56436734/181096883-4b1a428e-7d62-49f2-b479-5da157d73fed.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://docs.github.com/en/actions/creating-actions/creating-a-composite-action" target="_blank" rel="noopener nofollow ugc">Creating a composite action - GitHub Docs</a></h3>

  <p>In this guide, you'll learn how to build a composite action.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<p>
or</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.blog/2022-02-10-using-reusable-workflows-github-actions/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/56436734/181096884-f043d5f1-26b5-4317-a4be-a5102ec24e76.png" class="site-icon" width="32" height="32">

      <a href="https://github.blog/2022-02-10-using-reusable-workflows-github-actions/" target="_blank" rel="noopener nofollow ugc" title="05:56PM - 10 February 2022">The GitHub Blog – 10 Feb 22</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://user-images.githubusercontent.com/56436734/181096885-62111829-724a-4e9b-b78a-6dfbc534f093.png" class="thumbnail" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/7/f/7f4d1a62ee221a0a5f11fbd19874c4cfa03b35ef_2_690x362.png, https://aws1.discourse-cdn.com/github/optimized/3X/7/f/7f4d1a62ee221a0a5f11fbd19874c4cfa03b35ef_2_1035x543.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/7/f/7f4d1a62ee221a0a5f11fbd19874c4cfa03b35ef.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/7/f/7f4d1a62ee221a0a5f11fbd19874c4cfa03b35ef_2_10x10.png"></div>

<h3><a href="https://github.blog/2022-02-10-using-reusable-workflows-github-actions/" target="_blank" rel="noopener nofollow ugc">How to start using reusable workflows with GitHub Actions | The GitHub Blog</a></h3>

  <p>Reusable workflows offer a simple and powerful way to avoid copying and pasting workflows across your repositories.</p>

  <p>
    <span class="label1">Est. reading time: 6 minutes</span>
  </p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by ukreddy-erwin** [comment]
<p>I tried the steps in the second link composite action as is but.<br>
I created a repo in my account ukreddy-erwin, <a href="https://github.com/ukreddy-erwin/testactions" class="inline-onebox" rel="noopener nofollow ugc">GitHub - ukreddy-erwin/testactions</a></p>
<p>So, the workflow yaml changed as</p>
<pre><code class="lang-auto">
on: [push]

jobs:
  hello_world_job:
    runs-on: ubuntu-latest
    name: A job to say hello
    steps:
      - uses: actions/checkout@v3
      - id: foo
        uses: ukreddy-erwin/testactions@v1
        with:
          who-to-greet: 'Mona the Octocat'
      - run: echo random-number ${{ steps.foo.outputs.random-number }}
        shell: bash
</code></pre>
<p>But, it is failing as below error.</p>
<p><code>Error: Unable to resolve action </code>ukreddy-erwin/testactions@v1<code>, repository not found</code></p>
<p>So, I changed the repo to public and tried and got below error.</p>
<p><strong>How to keep them working in private repo?</strong></p>
<pre><code class="lang-auto">/home/runner/work/_temp/80546fcd-4e8c-4b37-8201-dfd471a8cd93.sh: line 1: /home/runner/work/_actions/ukreddy-erwin/testactions/v1/goodbye.sh: Permission denied

[18](https://github.com/ukreddy-erwin/testworkflow/runs/6653946764?check_suite_focus=true#step:3:23)Error: Process completed with exit code 126.
</code></pre>
<p>Even I added chmod +x command as one step in action but getting this,<br>
Run echo “/home/runner/work/_actions/ukreddy-erwin/testactions/ffbd787a4b9f0752347c5f62f2137460961e5a83” &gt;&gt; $GITHUB_PATH</p>
<p><a href="https://github.com/ukreddy-erwin/testworkflow/runs/6654053815?check_suite_focus=true#step:4:15" rel="noopener nofollow ugc">12</a> echo “/home/runner/work/_actions/ukreddy-erwin/testactions/ffbd787a4b9f0752347c5f62f2137460961e5a83” &gt;&gt; $GITHUB_PATH</p>
<p><a href="https://github.com/ukreddy-erwin/testworkflow/runs/6654053815?check_suite_focus=true#step:4:16" rel="noopener nofollow ugc">13</a> shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}</p>
<p><a href="https://github.com/ukreddy-erwin/testworkflow/runs/6654053815?check_suite_focus=true#step:4:18" rel="noopener nofollow ugc">14</a>Run chmod +x goodbye.sh</p>
<p><a href="https://github.com/ukreddy-erwin/testworkflow/runs/6654053815?check_suite_focus=true#step:4:19" rel="noopener nofollow ugc">15</a> chmod +x goodbye.sh</p>
<pre><code class="lang-auto">
[16](https://github.com/ukreddy-erwin/testworkflow/runs/6654053815?check_suite_focus=true#step:4:20) shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}

[17](https://github.com/ukreddy-erwin/testworkflow/runs/6654053815?check_suite_focus=true#step:4:22)chmod: cannot access 'goodbye.sh': No such file or directory

[18](https://github.com/ukreddy-erwin/testworkflow/runs/6654053815?check_suite_focus=true#step:4:23)Error: Process completed with exit code 1.
</code></pre>

**Comment by ViacheslavKudinov** [comment]
<p>Actions can be in internal repo (you need Enterprise GitHub org), but not private.<br>
You can checkout your action repo as any other private repo to specific subfolder and use it as “local” action.</p>
<p>So, unfortunately it is no easy way to have actions in private repo:s.</p>

**Comment by ukreddy-erwin** [comment]
<p>how to keep in the same repo as a sub folder and refer them?</p>

**Comment by ViacheslavKudinov** [comment]
<p>You can take a look at this thread <a href="https://github.community/t/path-to-action-in-the-same-repository-as-workflow/16952/2" class="inline-onebox">Path to action in the same repository as workflow - #2 by yanjingzhu</a></p>

**Comment by ViacheslavKudinov** [comment]
<p>Also this one is good example <a href="https://github.community/t/multiple-github-composite-actions-in-one-repo/199885/2" class="inline-onebox">Multiple github composite actions in one repo - #2 by mcaskill</a><br>
only token is needed to checkout private repository</p>

**Comment by ukreddy-erwin** [comment]
<p>Thanks <a class="mention" href="/viacheslavkudinov">@ViacheslavKudinov</a> , I tried the <a href="https://github.community/t/path-to-action-in-the-same-repository-as-workflow/16952/7">steps</a> but not working.</p>
<p>Below is the steps I tried.</p>
<p>I kept an action.yml in the root folder of repo and tried to reference it in the github workflow.<br>
But it is not able to find the yml file.<br>
I kept ls -la top to it and it shows the file but the workflow not picking it up.<br>
Can you see this once</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/ukreddy-erwin/testactions">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/ukreddy-erwin/testactions" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/21240709/181096893-76c58360-1dd0-478c-9f42-dfdc1608f558.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/f/afe0deed01c88308a06a1b7a3c48ee7e445fe251_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/a/f/afe0deed01c88308a06a1b7a3c48ee7e445fe251_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/a/f/afe0deed01c88308a06a1b7a3c48ee7e445fe251.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/f/afe0deed01c88308a06a1b7a3c48ee7e445fe251_2_10x10.png"></div>

<h3><a href="https://github.com/ukreddy-erwin/testactions" target="_blank" rel="noopener nofollow ugc">GitHub - ukreddy-erwin/testactions</a></h3>

  <p>Contribute to ukreddy-erwin/testactions development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by ViacheslavKudinov** [comment, marked as answer]
<p><a class="mention" href="/ukreddy-erwin">@ukreddy-erwin</a>, please, change step to</p>
<pre><code class="lang-auto">      - name: external step
        uses: ./
</code></pre>

**Comment by ukreddy-erwin** [comment]
<p>Thank you it works, so we it would pick from the folder directly.<br>
So, we can’t keep multiple yaml files in the same folder right?<br>
If can, how to refer in the users: ./</p>
<p>Also, i tried referring the input variable in the bash file of the action, but it is not expanding.Can you guide that also.</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/ukreddy-erwin/testactions/blob/8a9c78356cf770ba32016d03b56e4d3b2fd46ab6/goodbye.sh">
  <header class="source">

      <a href="https://github.com/ukreddy-erwin/testactions/blob/8a9c78356cf770ba32016d03b56e4d3b2fd46ab6/goodbye.sh" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/ukreddy-erwin/testactions/blob/8a9c78356cf770ba32016d03b56e4d3b2fd46ab6/goodbye.sh" target="_blank" rel="noopener nofollow ugc">ukreddy-erwin/testactions/blob/8a9c78356cf770ba32016d03b56e4d3b2fd46ab6/goodbye.sh</a></h4>


      <pre><code class="lang-sh">echo "Goodbye"
echo $userfulname</code></pre>




  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Also, is there anyway to keep these main workflow, actions, everything in a separate repository? so that these changes won’t add new commits to main codebase.</p>

**Comment by ViacheslavKudinov** [comment]
<aside class="quote no-group" data-username="ukreddy-erwin" data-post="10" data-topic="252984">
<div class="title">
<div class="quote-controls"></div>
 ukreddy-erwin:</div>
<blockquote>
<p>So, we can’t keep multiple yaml files in the same folder right?<br>
If can, how to refer in the users: ./</p>
</blockquote>
</aside>
<p>You can use subfolders if you wish and then it will be:</p>
<pre><code class="lang-auto">uses: subfolder
</code></pre>
<p>and it will pick up action.yaml from that subfolder.<br>
Moreover you can have several subfolders, depends on your use case.</p>
<aside class="quote no-group" data-username="ukreddy-erwin" data-post="10" data-topic="252984">
<div class="title">
<div class="quote-controls"></div>
 ukreddy-erwin:</div>
<blockquote>
<p>Also, is there anyway to keep these main workflow, actions, everything in a separate repository? so that these changes won’t add new commits to main codebase.</p>
</blockquote>
</aside>
<p>Yes, if you wish you can have one repo per one GitHub action and also utilise reusable workflow if you want to have workflow as separate code/repo.<br>
Repo for GHA has to be public for the simplest usage, or internal if you have Enterprise GH account/organisation.<br>
Please, read few articles <a href="https://docs.github.com/en/actions/using-workflows/reusing-workflows" class="inline-onebox" rel="noopener nofollow ugc">Reusing workflows - GitHub Docs</a><br>
<a href="https://docs.github.com/en/actions/using-workflows/sharing-workflows-secrets-and-runners-with-your-organization" class="inline-onebox" rel="noopener nofollow ugc">Sharing workflows, secrets, and runners with your organization - GitHub Docs</a></p>

**Comment by ukreddy-erwin** [comment]
<p>In jenkins, say I have my source code in repoA. I can keep my pipeline file in repoB and in a separate branch say testB.<br>
Then, in jenkins project, I can simply refer this repoB/testB branch pipeline file and it will run for repoA.</p>
<p>Can we set similarly here?</p>

**Comment by ViacheslavKudinov** [comment]
<p>It really depends what you want to achieve.<br>
But technically you can use GHA from any repo and branch/tag, just use<br>
<code>uses: my-org/my-action-repo@branch(or tag)</code>.<br>
Same for reusable workflow:<br>
<code>uses: my-org/my-gha-repo/.github/workflows/reusable-workflow.yml@branch</code></p>

**Comment by ViacheslavKudinov** [comment]
<p>If you don’t have Enterprise account/org it will be more complex (or maybe even not everything is possible) if you want to use private repositories for GHA:s and reusable workflows.</p>

**Comment by ukreddy-erwin** [comment]
<p>can we set something like.</p>
<p>Current repoA:</p>
<p>.github/workflows/main.yml</p>
<pre><code class="lang-auto">reference: myorg/repoB/branch1/workflow.yml
</code></pre>
<p>this workflow.yml is complete workflow, not an action.</p>

**Comment by ViacheslavKudinov** [comment]
<p>You have to use syntax eg:<br>
<code>uses: myorg/repoB/.github/workflows/workflow.yml@branch1</code><br>
part <code>.github/workflows</code> is mandatory as I know and it is where reusable workflow <strong>has to be</strong> in your <strong>repoB</strong>, branch name comes after <code>@</code>.</p>

**Comment by ukreddy-erwin** [comment]
<p>Can you answer the input variables part also.</p>
<p>It is not getting expanded in bash file.</p>
<p>tried referring the input variable in the bash file of the action, but it is not expanding.Can you guide that also.</p>
<h4>
<a name="ukreddy-erwintestactionsblob8a9c78356cf770ba32016d03b56e4d3b2fd46ab6goodbyesh-httpsgithubcomukreddy-erwintestactionsblob8a9c78356cf770ba32016d03b56e4d3b2fd46ab6goodbyesh-1" class="anchor" href="#ukreddy-erwintestactionsblob8a9c78356cf770ba32016d03b56e4d3b2fd46ab6goodbyesh-httpsgithubcomukreddy-erwintestactionsblob8a9c78356cf770ba32016d03b56e4d3b2fd46ab6goodbyesh-1"></a><a href="https://github.com/ukreddy-erwin/testactions/blob/8a9c78356cf770ba32016d03b56e4d3b2fd46ab6/goodbye.sh" rel="noopener nofollow ugc">ukreddy-erwin/testactions/blob/8a9c78356cf770ba32016d03b56e4d3b2fd46ab6/goodbye.sh </a>
</h4>

**Comment by ViacheslavKudinov** [comment]
<p><s>If I remember correctly inputs not available by default as environment variables.<br>
You need to use <code>env</code> section for job, eg</s><br>
<s>env:<br>
variable: ${{ inputs.name }}</s><br>
PS  IT WAS THE WRONG SUGGESTION, SORRY!!!</p>

**Comment by ukreddy-erwin** [comment]
<p>I tried and got below error.</p>
<aside class="onebox githubactions" data-onebox-src="https://github.com/ukreddy-erwin/testactions/actions/runs/2426296976">
  <header class="source">

      <a href="https://github.com/ukreddy-erwin/testactions/actions/runs/2426296976" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">


      <svg width="24" height="24" class="github-icon github-icon-failure" viewbox="0 0 24 24" aria-hidden="true"><path d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12zm8.036-4.024a.75.75 0 00-1.06 1.06L10.939 12l-2.963 2.963a.75.75 0 101.06 1.06L12 13.06l2.963 2.964a.75.75 0 001.061-1.06L13.061 12l2.963-2.964a.75.75 0 10-1.06-1.06L12 10.939 9.036 7.976z"></path></svg>

  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/ukreddy-erwin/testactions/actions/runs/2426296976" target="_blank" rel="noopener nofollow ugc">Update goodbye.sh</a>
    </h4>

      <div class="github-info">
        CI <span class="github-run-number">#13</span>
      </div>
  </div>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by ViacheslavKudinov** [comment]
<p>Let me double-check and I will be back soon.</p>
